### PR TITLE
refactor: use new Read Buffer chunk-oriented API

### LINKED ISSUE #1130: Error creating scan for table http_api_requests_total chunk Read Buffer Error in chunk chunk id does not exist
```
alamb@ip-10-0-0-124 influxdb_iox % ./target/debug/influxdb_iox --host http://localhost:1234 database query 844910ece80be8bc_7be09b71c487d5d3 "select count(*) from http_api_requests_total;"
./target/debug/influxdb_iox --host http://localhost:1234 database query 844910ece80be8bc_7be09b71c487d5d3 "select count(*) from http_api_requests_total;"
Error querying: status: Internal, message: "Internal error reading points from database 844910ece80be8bc_7be09b71c487d5d3:  Arrow error: External error: Execution error: Error creating scan for table http_api_requests_total chunk 146: Read Buffer Error in chunk 146: chunk id does not exist: 146", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Tue, 06 Apr 2021 16:31:32 GMT"} }
alamb@ip-10-0-0-124 influxdb_iox % 
```

__Steps to reproduce:__
Try to query a database that has data being written to it actively and that the chunk lifecycle rules cause data to get dropped out of the read buffer

The query will work, but occasionally it will fail with the error above

__Expected behaviour:__
The query should complete successfully

__Actual behaviour:__

Error : "Error creating scan for table http_api_requests_total chunk 146: Read Buffer Error in chunk 146: chunk id does not exist: 146","

__Environment info:__
internal tools cluster

### PR DESCRIPTION
**Pull request by e-dard**
Closes #1130 (probably)

This PR leverages the new public API that the Read Buffer exposes, which is now at the chunk level to more easily integrate with design choices in the query engine / catalog. 

- [x] expose new public API in Read Buffer at Chunk level;
- [x] use new Read Buffer Chunk API (this PR); 
- [ ] remove deprecated Read Buffer API/code.

### CODE REVIEW

File: server/src/db.rs [hunk 1]
```diff
@@ -326,36 +320,22 @@ impl Db {
...
-                    return DropMovingChunk {
-                        partition_key,
-                        chunk_id,
-                        chunk_state,
-                    }
-                    .fail()
-                }
-                ChunkState::Moved(_) => {
-                    drop_rb = true;
+            ensure!(
```
Comment (e-dard) [inline server/src/db.rs:323]: I think that this is all that needs to happen now.

File: server/src/db.rs [hunk 2]
```diff
@@ -405,6 +385,10 @@ impl Db {
         let mut batches = Vec::new();
         let table_stats = mb_chunk.table_summaries();
 
+        // create a new read buffer chunk.
+        let rb_chunk = ReadBufferChunk::new(chunk_id);
```
Comment (e-dard) [inline server/src/db.rs:389]: This is effectively how we make "a new read buffer" each time we need a new read buffer chunk.

File: server/src/db/catalog/chunk.rs [hunk 3]
```diff
@@ -32,13 +32,13 @@ pub enum ChunkState {
     Moving(Arc<MBChunk>),
 
     /// Chunk has been completely loaded in the read buffer
-    Moved(Arc<ReadBufferDb>), // todo use read buffer chunk instead of ReadBufferDb
+    Moved(Arc<ReadBufferChunk>),
```
Comment (e-dard) [inline server/src/db/catalog/chunk.rs:35]: Now the catalog just uses read buffer chunks.

**Comment by e-dard** [comment]
Oh and a big shout out to @carols10cents. I went with another direction on returning `impl Iterator<Item = RecordBatch>` from the Read Buffer's `read_filter` method and it turns out that that is ALL KINDS OF FUTURES STREAMS MADNESS. So after hours of us banging our heads against an coding wall we gave up and I just returned a concrete results type. 😞

File: server/src/db.rs [hunk 4]
```diff
@@ -205,13 +205,17 @@ const STARTING_SEQUENCE: u64 = 1;
...
+    /// depending on their stage in the data lifecycle. Currently there are
+    /// three backing engines for Chunks:
+    ///
+    ///  - The Mutable Buffer where chunks are mutable but also queryable;
+    ///  - The Read Buffer where chunks are immutable and stored in an optimised
+    ///    compressed form for small footprint and fast query execution; and
+    ///  - The Parquet Buffer where chunks are backed by Parquet file data.
     catalog: Arc<Catalog>,
 
-    /// The read buffer holds chunk data in an in-memory optimized
```
Comment (alamb) [inline server/src/db.rs:211]: 🎉

File: server/src/db/catalog/chunk.rs [hunk 5]
```diff
@@ -202,14 +196,32 @@ impl Chunk {
             ChunkState::Invalid => {}
             ChunkState::Open(chunk) | ChunkState::Closing(chunk) => chunk.all_table_names(names),
             ChunkState::Moving(chunk) => chunk.all_table_names(names),
-            ChunkState::Moved(db) => {
-                db.all_table_names(self.partition_key.as_str(), &[self.id], names)
+            ChunkState::Moved(chunk) => {
+                // TODO - the RB API returns a new set each time, so maybe this
+                // method should be updated to do the same across the mutable
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:201]: Yeah I think harmonizing the read buffer's API and the other chunk APIs is a good idea. 👍

**Review by alamb — APPROVED** [review]
This is a wonderful PR. Thank you @e-dard

**Pull request merged**